lib/checkstl.cpp

```diff
@@ -1941,18 +1941,32 @@ void CheckStl::string_c_str()
             if (Token::Match(tok, "throw %var% . c_str|data ( ) ;") && isLocal(tok->next()) &&
                 tok->next()->variable() && tok->next()->variable()->isStlStringType()) {
                 string_c_strThrowError(tok);
-            } else if (tok->variable() && tok->strAt(1) == "=") {
-                if (Token::Match(tok->tokAt(2), "%var% . str ( ) . c_str|data ( ) ;")) {
-                    const Variable* var = tok->variable();
-                    const Variable* var2 = tok->tokAt(2)->variable();
-                    if (var->isPointer() && var2 && var2->isStlType(stl_string_stream))
-                        string_c_strError(tok);
-                } else if (Token::Match(tok->tokAt(2), "%name% (") &&
-                           Token::Match(tok->linkAt(3), ") . c_str|data ( ) ;") &&
-                           tok->tokAt(2)->function() && Token::Match(tok->tokAt(2)->function()->retDef, "std :: string|wstring %name%")) {
-                    const Variable* var = tok->variable();
-                    if (var->isPointer())
+            } else if (tok->variable() && tok->next()->isAssignmentOp() && tok->next()->astOperand2()) {
+                const Token* op2 = tok->next()->astOperand2(); // a = a.str().c_str();
+                const Token* op21 = op2->astOperand1();
+                if (op2->str() != "(" || !op21 || op21->str() != ".")
+                    continue;
+
+                if (op21->strAt(1) != "c_str" && op21->strAt(1) != "data")
+                    continue;
+
+                const Token* op211 = op21->astOperand1();
+                if (Token::Match(op21->tokAt(-5), "%var% . str|data ( )")) { // a = a.str().c_str();
```

Comment:
hmm.. could this be extended to `expr.str().c_str()` and check if `expr` result is some std container? for instance `a[0].str().c_str()`